### ISSUE TITLE
[feat]: CLI support for package.json subpath imports

### ISSUE THREAD

**Issue by jeremy-code** [shadcn-ui/ui#2960]
### Feature description

In TypeScript 5.4, [subpath imports](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#auto-import-support-for-subpath-imports) are now supported. In `components.json`, if the subpath import is added as an alias like:

```json
"aliases": {
    "components": "#components"
}
```

It will create a new folder called #components, and put all the components in there. Conversely, if the alias is set to the actual components directory (e.g. ./src/components/ui), an error will be thrown as an array of errors like:

```
{
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "resolvedPaths",
      "components"
    ],
    "message": "Required"
},
```

Not setting any alias at all results in error "Invalid configuration found in ../components.json".

In the meantime, I have set it with the subpath imports as the alias, and added bash commands to move it to the correctly place, but this isn't ideal. 

```
"ui:add": "yarn dlx shadcn-ui@latest add && mv ./#components/ui/* ./src/components/ui/ && rm -rf ./#components",
```

### Before submitting

- [X] I've made research efforts and searched the documentation
- [X] I've searched for existing issues and PRs

**Comment by shadcn** [comment]
This issue has been automatically closed because it received no activity for a while. If you think it was closed by accident, please leave a comment. Thank you.

**Comment by jam-fran** [comment]
Turborepo now officially recommends node subpath imports for internal packages in [their docs](https://turbo.build/repo/docs/core-concepts/internal-packages):

> _A library that is being transpiled by its consumer cannot use the compilerOptions.paths configuration because TypeScript assumes that source code is being transpiled in the package where it is written. If you're using Typescript 5.4 or later, we recommend [using Node.js subpath imports](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#auto-import-support-for-subpath-imports)._

Thanks @jeremy-code for the workaround, but would be amazing if shadcn supported this at some point, since I imagine this will become much more prevalent in the near future.

**Comment by sqmasep** [comment]
I'm facing the same issue, it would be cool that the CLI get fixed on the path alias. Currently, it seems like if the CLI detects that in the tsconfig there is a "wrong" alias (such as `#utils`), it fallbacks to `#/lib/utils.` 

Node subpath can only begin with a `#` so it doesn't get confused with public scoped packages (that start with `@`)
The problem is that node subpath doesn't allow an alias that starts with `#/`. And shadcn CLI doesn't seem to work with anything that don't start with `<character>/`. This trailing slash makes the problem, basically

The only way to fix it is whenever you add a new component with the CLI, you need to fix the path alias by hand (unless someone else finds another solution?)

**Comment by SanderCokart** [comment]
Reactivate this please!

**Comment by jeremy-code** [comment]
A workaround that involves not using aliases is using [Node.js self-referencing](https://nodejs.org/api/packages.html#self-referencing-a-package-using-its-name).

For instance, see the following code:

```jsonc
// package.json
{
  "name": "@repo/ui",
  "exports": {
    "./globals.css": "./globals.css",
    "./postcss.config": "./postcss.config.js",
    "./tailwind.config": "./tailwind.config.ts",
    "./utils": "./utils.ts",
    "./components/*": "./components/*.tsx",
    "./components/ui/*": "./components/ui/*.tsx"
  },
}

// components.json
{
  "aliases": {
    "utils": "@repo/ui/utils",
    "components": "@repo/ui/components"
  }
}
```

Then, internally (in @repo/ui), imports will look like `import { cn } from "@repo/ui/utils";` and in external apps like `import { Switch } from "@repo/ui/components/ui/switch";`. If you're not using a module bundler, you'll have to update the export map for `.js`.

IMO, due to the package structure (few imports), this is more useful than using aliasing since it just bases everything based on what you export. The only issue is if you have some internal tool in the library that you don't want to export, but at least for `shadcn/ui` CLI, there isn't any instances of this.

**Comment by bdbergeron** [comment]
After playing around with things for a while, I was able to get subpath imports working with shadcn/ui components _nearly_ perfectly. My package is a [Compiled Package](https://turbo.build/repo/docs/core-concepts/internal-packages#compiled-packages), thus the `dist` output path being referenced, but I believe this should work if yours is a [Just-In-Time Package](https://turbo.build/repo/docs/core-concepts/internal-packages#just-in-time-packages) if you update the `imports` path to point to your TS files instead of the compiled files.

```
// package.json
{
  ...
  "imports": {
    "#*": "./dist/*.js"
  },
  ...
}
```

```
// tsconfig.json
{
  ...
  "compilerOptions": {
    "paths": {
      "#*": ["./src/*"]
    },
  }
  ...
}
```

```
// components.json
{
  ...
  "aliases": {
    "components": "#components",
    "hooks": "#hooks",
    "lib": "#lib",
    "ui": "#components/ui",
    "utils": "#lib/utils"
  }
  ...
}
```

It is unfortunate to need the TS path alias in addition to the subpath import, but this gets you 99% there. The only thing that **_does not_** seem to work with this setup is the generated `cn` import, which for some reason gets imported as `#/lib/utils` instead of `#lib/utils`. As @sqmasep pointed out above, you'll have to fix this import by hand.

**Comment by jdfm** [comment]
> A workaround that involves not using aliases is using [Node.js self-referencing](https://nodejs.org/api/packages.html#self-referencing-a-package-using-its-name).
> 
> For instance, see the following code:
> 
> ```js
> // package.json
> {
>   "name": "@repo/ui",
>   "exports": {
>     "./globals.css": "./globals.css",
>     "./postcss.config": "./postcss.config.js",
>     "./tailwind.config": "./tailwind.config.ts",
>     "./utils": "./utils.ts",
>     "./components/*": "./components/*.tsx",
>     "./components/ui/*": "./components/ui/*.tsx"
>   },
> }
> 
> // components.json
> {
>   "aliases": {
>     "utils": "@repo/ui/utils",
>     "components": "@repo/ui/components"
>   }
> }
> ```
> 
> Then, internally (in @repo/ui), imports will look like `import { cn } from "@repo/ui/utils";` and in external apps like `import { Switch } from "@repo/ui/components/ui/switch";`. If you're not using a module bundler, you'll have to update the export map for `.js`.
> 
> IMO, due to the package structure (few imports), this is more useful than using aliasing since it just bases everything based on what you export. The only issue is if you have some internal tool in the library that you don't want to export, but at least for `shadcn/ui` CLI, there isn't any instances of this.

I ended up going the self-referencing route, here's what the exports (in package.json) ended up looking like:

```json
  "exports": {
    "./utils/*": {
      "types": "./src/lib/utils/*.ts",
      "default": "./src/lib/utils/*.ts"
    },
    "./lib/*": {
      "types": "./src/lib/*.ts",
      "default": "./src/lib/*.ts"
    },
    "./components/*": {
      "types": "./src/components/*.tsx",
      "default": "./src/components/*.tsx"
    },
    "./ui/*": {
      "types": "./src/components/ui/*.tsx",
      "default": "./src/components/ui/*.tsx"
    },
    "./hooks/*": {
      "types": "./src/hooks/*.tsx",
      "default": "./src/hooks/*.tsx"
    }
  }
```

And in tsconfig.json:

```json
    "paths": {
      "@repo/ui/*": ["./src/*"]
    }
```

This helped me get rid of `.ts/tsx` extensions at the app level when importing code, allows bundlers and typescript to properly identify what code is relevant, and allows shadcn cli to work (for the most part, the bugs that exist around the utils import at the moment still affect us here). This also helped me keep this library configured as an "internal" typescript package as outlined in https://turbo.build/blog/you-might-not-need-typescript-project-references#internal-typescript-packages

I think I ran into issues with `transpilePackages` config in nextjs when I tried to use the `"imports"` setup in package.json and tried to build, not entirely sure as it's been a while since I tested that particular setup.

**Comment by shadcn** [comment]
This issue has been automatically marked as stale due to two years of inactivity. It will be closed in 7 days unless there’s further input. If you believe this issue is still relevant, please leave a comment or provide updated details. Thank you.

**Comment by dpnova** [comment]
Agreed - using the CLI is a big pain in our projects that use subpath imports

**Comment by shadcn** [comment]
This issue has been automatically closed due to one year of inactivity. If you’re still experiencing a similar problem or have additional details to share, please open a new issue following our current issue template. Your updated report helps us investigate and address concerns more efficiently. Thank you for your understanding! (This is an automated message)

**Comment by valerii15298** [comment]
Reactivate this please!

**Comment by kentcdodds** [comment]
We recently removed all path aliases from [Epic Stack's](https://github.com/epicweb-dev/epic-stack) default `tsconfig` (see [epicweb-dev/epic-stack#1060](https://github.com/epicweb-dev/epic-stack/issues/1060) for the fallout) so new projects rely on TypeScript 5.4+ subpath imports like `#app`.

When we run `npx shadcn@latest add dropdown-menu`, the CLI resolves the alias literally and scaffolds into `#app/components/ui` instead of the intended `./app/components/ui`.

Because the CLI doesn’t understand the `#` subpath, teams have to add manual shell steps to move the files after every install. Supporting package.json subpath aliases (or at least ignoring the leading `#`) would let us keep the cleaner compiler defaults without breaking the install workflow.

**Comment by shadcn** [comment]
I just got the ping on this. Reopening...

**Comment by shadcn** [comment]
I don't know if I can get to it this week. But I can review and merge a PR if anyone want to take a look. Thank you.

**Comment by mmikhan** [comment]
Facing this as well. Needs this badly 🙏

**Comment by aerojeyenth** [comment]
+1

**Comment by AdeboyeDN** [comment]
> I don't know if I can get to it this week. But I can review and merge a PR if anyone want to take a look. Thank you.

I can take this up

**Comment by bobfw** [comment]
If this functionality is implemented, shadcn will be much simpler on monorepo.

**Comment by gkurt** [comment]
Created a PR for this at #9759

**Comment by shadcn** [comment]
@KurtGokhan Thanks for the PR. I've been looking into this as well. I'm doing some major work on how paths are resolved right now. I'll take a look at your PR and see how we can make this work. Thank you.

**Comment by firxworx** [comment]
@shadcn do you think the whole setup around paths is trying to be "too smart for its own good" and thus not intuitive and not behaving the way users expect, and running into frequent future-proofing / edge case issues?

**Declarative & Explicit**

If devs could spell out: 

- _where_ they wanted the cli to create files (e.g. absolute or relative path from `components.json` or perhaps an optional base)
- _what_ import paths/patterns to use

Then the CLI could be "dumb" and do simple string interpolation / concatenation from there.

It would be simple to reason about and there'd be no baked in assumptions or hidden behaviours that could lead to unexpected or undesirable outcomes.

Everything would have worked in a monorepo, with subpath imports, etc. from the start.

It'd still be possible to have smart fallbacks so the config can be tight, however the power to simply spell everything out when required would be amazing.

You could still keep the old `aliases` for backwards compatibility and simply introduce a new object.

**Descriptive & Self-Documenting**

FWIW I also think the alias property names have always been confusing and I recall a big source of frustration back when I first tried shadcn/ui when it first launched. I'm sure your own understanding as evolved since the earliest iterations!

Personally I found relationship between them wasn't clear and IIRC it has slightly changed over time e.g.  "lib" + "utils"; "components" + "ui", etc. 

Since the `cn()` function is so central to shadcn/ui I suggest having import for the classname merge function to be its own config property, which would also be flexible for customizing tailwind-merge and other more custom setups. 

With design systems and customized tailwind that can be such a special function that its not simply just another util. 

Thanks for your time, shadcn/ui was a game changer for so many of us :)

**Comment by shadcn** [comment]
@firxworx Agreed. I've been looking into this as well (I have a working POC that I planned to ship in v4 but didn't make it). My plan right now is to get subpath imports working. Then look into this again. (There are other things I want to support as well like file naming...)

Actively looking into making this simpler.

**Comment by shadcn** [comment]
Shipped in `canary` (finally). Help me test: `npx shadcn@canary init` #10070 

(Kudos to @KurtGokhan for starting the work on this. I took a different approach to their PR but it gave me some ideas where to look. Thank you. Appreciate it)

**Comment by bobfw** [comment]
Hi @shadcn Thank you for your efforts. When will this feature be released?

**Comment by kasperpeulen** [comment]
Would love to start using this!
